File: backup_manager.py
from __future__ import annotations

import fnmatch
import hashlib
import json
import os
import shutil
import tempfile
import zipfile
from datetime import datetime, timezone
from pathlib import Path


class BackupManager:
    BACKUP_VERSION = 1
    EXCLUDE_GLOBS = ["*.key", "*token*", "secrets.json"]
    ROOT_MUTABLE_FILES = {
        "settings.json": "config/settings.json",
        "pins.json": "data_root/pins.json",
        "tours.json": "data_root/tours.json",
        "geocode_cache.json": "data_root/geocode_cache.json",
        "config.json": "data_root/config.json",
    }
    RESTORE_LABELS = {
        "orders": "Aufträge & Adressen",
        "tours": "Liefertouren",
        "employees": "Mitarbeiter",
        "vehicles": "Fahrzeuge",
        "settings": "Einstellungen",
        "misc": "Zusatzdaten",
        "other_data": "Weitere Daten",
    }

    def __init__(self, app_name: str, config_dir: Path, data_dir: Path, backup_dir: Path):
        self.app_name = str(app_name or "App").strip() or "App"
        self.config_dir = Path(config_dir)
        self.data_dir = Path(data_dir)
        self.backup_dir = Path(backup_dir)
        self.backup_dir.mkdir(parents=True, exist_ok=True)

    def create_backup(self, mode: str) -> Path:
        mode = str(mode or "full").strip().lower()
        if mode == "incremental":
            return self.create_incremental_backup()
        return self.create_full_backup()

    def create_full_backup(self) -> Path:
        self.backup_dir.mkdir(parents=True, exist_ok=True)
        timestamp = datetime.now().strftime("%Y-%m-%d_%H%M%S")
        target = self.backup_dir / f"{self.app_name}_backup_FULL_{timestamp}.bak"
        file_index, file_map, skipped = self.compute_file_index()
        manifest = self.build_manifest(file_index=file_index, mode="full", base_backup=None, deleted_paths=None)

        with tempfile.NamedTemporaryFile(delete=False, suffix=".bak", dir=str(self.backup_dir)) as tmp:
            tmp_path = Path(tmp.name)

        try:
            with zipfile.ZipFile(tmp_path, "w", compression=zipfile.ZIP_DEFLATED) as archive:
                self._write_snapshot_entries(archive=archive, file_map=file_map)
                archive.writestr("manifest.json", json.dumps(manifest, ensure_ascii=False, indent=2))
                archive.writestr("meta/log.txt", self._build_log(skipped))
            os.replace(tmp_path, target)
        finally:
            if tmp_path.exists():
                try:
                    tmp_path.unlink()
                except Exception:
                    pass
        return target

    def create_incremental_backup(self) -> Path:
        latest = self.find_latest_backup()
        if latest is None:
            return self.create_full_backup()

        try:
            previous_manifest = self._read_manifest(latest)
            previous_index = {
                entry["path"]: entry
                for entry in previous_manifest.get("file_index", [])
                if isinstance(entry, dict) and entry.get("path")
            }
        except Exception:
            return self.create_full_backup()

        file_index, file_map, skipped = self.compute_file_index()
        current_index = {entry["path"]: entry for entry in file_index}

        changed_paths = {
            path for path, entry in current_index.items() if previous_index.get(path, {}).get("sha256") != entry.get("sha256")
        }
        deleted_paths = sorted(path for path in previous_index if path not in current_index)

        with tempfile.NamedTemporaryFile(delete=False, suffix=".bak", dir=str(self.backup_dir)) as tmp:
            tmp_path = Path(tmp.name)

        try:
            with zipfile.ZipFile(latest, "r") as source_archive, zipfile.ZipFile(
                tmp_path, "w", compression=zipfile.ZIP_DEFLATED
            ) as target_archive:
                self._copy_zip_without_meta(source_archive, target_archive, skip_paths=changed_paths | {"manifest.json", "meta/log.txt"})
                self._write_snapshot_entries(target_archive, file_map, include_paths=changed_paths)
                manifest = self.build_manifest(
                    file_index=file_index,
                    mode="incremental",
                    base_backup=latest.name,
                    deleted_paths=deleted_paths,
                )
                target_archive.writestr("manifest.json", json.dumps(manifest, ensure_ascii=False, indent=2))
                target_archive.writestr("meta/log.txt", self._build_log(skipped))
            os.replace(tmp_path, latest)
        finally:
            if tmp_path.exists():
                try:
                    tmp_path.unlink()
                except Exception:
                    pass
        return latest

    def find_latest_backup(self) -> Path | None:
        backups = sorted(self.backup_dir.glob("*.bak"), key=lambda item: item.stat().st_mtime, reverse=True)
        return backups[0] if backups else None

    def cleanup_old_backups(self, retention_days: int) -> None:
        try:
            retention_days = int(retention_days)
        except Exception:
            return
        cutoff = datetime.now(timezone.utc).timestamp() - (retention_days * 86400)
        for bak_file in self.backup_dir.glob("*.bak"):
            try:
                if bak_file.stat().st_mtime < cutoff:
                    bak_file.unlink()
            except Exception:
                continue

    def build_manifest(self, file_index, mode, base_backup=None, deleted_paths=None) -> dict:
        return {
            "backup_version": self.BACKUP_VERSION,
            "created_at_iso": datetime.now(timezone.utc).isoformat(),
            "backup_type": mode,
            "app_name": self.app_name,
            "source_paths": {
                "config_dir": str(self.config_dir),
                "data_dir": str(self.data_dir),
            },
            "file_index": list(file_index or []),
            "deleted_paths": list(deleted_paths or []),
            "incremental_base": base_backup,
        }

    def compute_file_index(self):
        file_index = []
        file_map = {}
        skipped = []
        for archive_path, fs_path in self.scan_files():
            try:
                stat = fs_path.stat()
                digest = self._sha256(fs_path)
            except PermissionError:
                skipped.append(f"Permission denied: {fs_path}")
                continue
            except Exception as exc:
                skipped.append(f"Skipped {fs_path}: {exc}")
                continue

            file_index.append(
                {
                    "path": archive_path,
                    "size": int(stat.st_size),
                    "mtime": float(stat.st_mtime),
                    "sha256": digest,
                }
            )
            file_map[archive_path] = fs_path

        file_index.sort(key=lambda item: item["path"])
        return file_index, file_map, skipped

    def scan_files(self):
        # Back up mutable root-level app files as well because pins/tours live next to main.py.
        for file_name, archive_path in self.ROOT_MUTABLE_FILES.items():
            source_path = self.config_dir / file_name
            if source_path.exists() and self._is_included(source_path.name):
                yield archive_path, source_path

        if self.data_dir.exists():
            for path in sorted(self.data_dir.rglob("*")):
                if not path.is_file():
                    continue
                relative = path.relative_to(self.data_dir).as_posix()
                if not self._is_included(relative):
                    continue
                yield f"data/{relative}", path

    def restore_backup(self, bak_path: Path, target_data_dir: Path, target_config_dir: Path, selected_groups=None) -> None:
        bak_path = Path(bak_path)
        target_data_dir = Path(target_data_dir)
        target_config_dir = Path(target_config_dir)
        target_data_dir.mkdir(parents=True, exist_ok=True)
        target_config_dir.mkdir(parents=True, exist_ok=True)
        allowed_groups = None if not selected_groups or "all" in selected_groups else set(selected_groups)

        with zipfile.ZipFile(bak_path, "r") as archive:
            manifest = self._read_manifest(bak_path)
            for member in archive.namelist():
                if member in {"manifest.json", "meta/log.txt"} or member.endswith("/"):
                    continue
                if allowed_groups is not None and self._classify_archive_member(member) not in allowed_groups:
                    continue
                if member == "config/settings.json":
                    destination = target_config_dir / "settings.json"
                elif member.startswith("data_root/"):
                    destination = target_config_dir / Path(member).relative_to("data_root")
                elif member.startswith("data/"):
                    destination = target_data_dir / Path(member).relative_to("data")
                else:
                    continue
                destination.parent.mkdir(parents=True, exist_ok=True)
                with archive.open(member, "r") as source, destination.open("wb") as target:
                    shutil.copyfileobj(source, target)

            for deleted in manifest.get("deleted_paths", []) or []:
                if allowed_groups is not None and self._classify_archive_member(str(deleted)) not in allowed_groups:
                    continue
                if str(deleted).startswith("data_root/"):
                    destination = target_config_dir / Path(deleted).relative_to("data_root")
                elif str(deleted).startswith("data/"):
                    destination = target_data_dir / Path(deleted).relative_to("data")
                else:
                    continue
                if destination.exists():
                    try:
                        destination.unlink()
                    except Exception:
                        pass

    def _copy_zip_without_meta(self, source_archive, target_archive, skip_paths):
        skipped = set(skip_paths or set())
        copied = set()
        for info in source_archive.infolist():
            if info.filename in skipped or info.filename in copied:
                continue
            copied.add(info.filename)
            target_archive.writestr(info, source_archive.read(info.filename))

    def _write_snapshot_entries(self, archive, file_map, include_paths=None):
        include = set(include_paths or file_map.keys())
        for archive_path in sorted(include):
            source_path = file_map.get(archive_path)
            if source_path is None or not source_path.exists():
                continue
            archive.write(source_path, arcname=archive_path)

    def _read_manifest(self, bak_path: Path) -> dict:
        with zipfile.ZipFile(bak_path, "r") as archive:
            with archive.open("manifest.json", "r") as handle:
                return json.load(handle)

    def _build_log(self, skipped_messages):
        lines = [
            f"Backup created at {datetime.now(timezone.utc).isoformat()}",
        ]
        for message in skipped_messages or []:
            lines.append(message)
        return "\n".join(lines) + "\n"

    def _is_included(self, relative_path: str) -> bool:
        name = str(relative_path or "")
        return not any(fnmatch.fnmatch(name, pattern) for pattern in self.EXCLUDE_GLOBS)

    def _classify_archive_member(self, member: str) -> str:
        path = str(member or "").replace("\\", "/")
        if path == "config/settings.json":
            return "settings"
        if path == "data_root/pins.json":
            return "orders"
        if path == "data_root/tours.json":
            return "tours"
        if path == "data_root/geocode_cache.json" or path == "data_root/config.json":
            return "misc"
        if path == "data/employees.json":
            return "employees"
        if path == "data/vehicles.json":
            return "vehicles"
        if path.startswith("data/"):
            return "other_data"
        return "misc"

    def _sha256(self, path: Path) -> str:
        digest = hashlib.sha256()
        with path.open("rb") as handle:
            while True:
                chunk = handle.read(1024 * 1024)
                if not chunk:
                    break
                digest.update(chunk)
        return digest.hexdigest()

File: Tourenplaner.CSharp/src/Tourenplaner.CSharp.Infrastructure/Backups/BackupManager.cs
using System.IO.Compression;
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Serialization;
using Tourenplaner.CSharp.Domain.Enums;

namespace Tourenplaner.CSharp.Infrastructure.Backups;

public sealed class BackupManager
{
    private const int BackupVersion = 1;
    private static readonly string[] ExcludeGlobs = ["*.key", "*token*", "secrets.json"];

    private static readonly IReadOnlyDictionary<string, string> RootMutableFiles = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
    {
        ["settings.json"] = "config/settings.json",
        ["pins.json"] = "data_root/pins.json",
        ["tours.json"] = "data_root/tours.json",
        ["geocode_cache.json"] = "data_root/geocode_cache.json",
        ["config.json"] = "data_root/config.json",
    };

    public static IReadOnlyDictionary<string, string> RestoreLabels { get; } = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
    {
        ["orders"] = "Aufträge & Adressen",
        ["tours"] = "Liefertouren",
        ["employees"] = "Mitarbeiter",
        ["vehicles"] = "Fahrzeuge",
        ["settings"] = "Einstellungen",
        ["misc"] = "Zusatzdaten",
        ["other_data"] = "Weitere Daten",
    };

    private readonly string _appName;
    private readonly DirectoryInfo _configDirectory;
    private readonly DirectoryInfo _dataDirectory;
    private readonly DirectoryInfo _backupDirectory;

    public BackupManager(string appName, string configDirectory, string dataDirectory, string backupDirectory)
    {
        _appName = string.IsNullOrWhiteSpace(appName) ? "App" : appName.Trim();
        _configDirectory = new DirectoryInfo(configDirectory);
        _dataDirectory = new DirectoryInfo(dataDirectory);
        _backupDirectory = new DirectoryInfo(backupDirectory);
        _backupDirectory.Create();
    }

    public Task<string> CreateBackupAsync(BackupMode mode, CancellationToken cancellationToken = default)
        => Task.Run(() => CreateBackup(mode, cancellationToken), cancellationToken);

    public string CreateBackup(BackupMode mode, CancellationToken cancellationToken = default)
        => mode == BackupMode.Incremental ? CreateIncrementalBackup(cancellationToken) : CreateFullBackup(cancellationToken);

    public string? FindLatestBackup()
        => _backupDirectory.Exists
            ? _backupDirectory.EnumerateFiles("*.bak", SearchOption.TopDirectoryOnly).OrderByDescending(file => file.LastWriteTimeUtc).FirstOrDefault()?.FullName
            : null;

    public void CleanupOldBackups(int retentionDays)
    {
        var cutoff = DateTimeOffset.UtcNow.AddDays(-Math.Max(1, retentionDays));
        if (!_backupDirectory.Exists)
        {
            return;
        }

        foreach (var file in _backupDirectory.EnumerateFiles("*.bak", SearchOption.TopDirectoryOnly))
        {
            if (file.LastWriteTimeUtc < cutoff.UtcDateTime)
            {
                file.Delete();
            }
        }
    }

    public Task RestoreBackupAsync(string backupPath, string targetDataDirectory, string targetConfigDirectory, IReadOnlyCollection<string>? selectedGroups, CancellationToken cancellationToken = default)
        => Task.Run(() => RestoreBackup(backupPath, targetDataDirectory, targetConfigDirectory, selectedGroups, cancellationToken), cancellationToken);

    public void RestoreBackup(string backupPath, string targetDataDirectory, string targetConfigDirectory, IReadOnlyCollection<string>? selectedGroups, CancellationToken cancellationToken = default)
    {
        var allowedGroups = selectedGroups is null || selectedGroups.Count == 0 || selectedGroups.Contains("all", StringComparer.OrdinalIgnoreCase)
            ? null
            : new HashSet<string>(selectedGroups, StringComparer.OrdinalIgnoreCase);

        var targetData = new DirectoryInfo(targetDataDirectory);
        var targetConfig = new DirectoryInfo(targetConfigDirectory);
        targetData.Create();
        targetConfig.Create();

        using var archive = ZipFile.OpenRead(backupPath);
        var manifest = ReadManifest(archive);
        foreach (var entry in archive.Entries)
        {
            cancellationToken.ThrowIfCancellationRequested();
            if (string.Equals(entry.FullName, "manifest.json", StringComparison.OrdinalIgnoreCase)
                || string.Equals(entry.FullName, "meta/log.txt", StringComparison.OrdinalIgnoreCase)
                || string.IsNullOrWhiteSpace(entry.Name))
            {
                continue;
            }

            if (allowedGroups is not null && !allowedGroups.Contains(ClassifyArchiveMember(entry.FullName)))
            {
                continue;
            }

            var destination = ResolveRestorePath(entry.FullName, targetData.FullName, targetConfig.FullName);
            if (destination is null)
            {
                continue;
            }

            Directory.CreateDirectory(Path.GetDirectoryName(destination)!);
            entry.ExtractToFile(destination, overwrite: true);
        }

        foreach (var deletedPath in manifest.DeletedPaths)
        {
            cancellationToken.ThrowIfCancellationRequested();
            if (allowedGroups is not null && !allowedGroups.Contains(ClassifyArchiveMember(deletedPath)))
            {
                continue;
            }

            var destination = ResolveRestorePath(deletedPath, targetData.FullName, targetConfig.FullName);
            if (!string.IsNullOrWhiteSpace(destination) && File.Exists(destination))
            {
                File.Delete(destination);
            }
        }
    }

    private string CreateFullBackup(CancellationToken cancellationToken)
    {
        _backupDirectory.Create();
        var timestamp = DateTime.Now.ToString("yyyy-MM-dd_HHmmss");
        var target = Path.Combine(_backupDirectory.FullName, $"{_appName}_backup_FULL_{timestamp}.bak");
        var snapshot = ComputeFileSnapshot(cancellationToken);
        var manifest = BuildManifest(snapshot.Index, "full", null, []);

        using var archive = ZipFile.Open(target, ZipArchiveMode.Create);
        WriteSnapshotEntries(archive, snapshot.Map, snapshot.Map.Keys, cancellationToken);
        WriteTextEntry(archive, "manifest.json", JsonSerializer.Serialize(manifest, JsonOptions));
        WriteTextEntry(archive, "meta/log.txt", BuildLog(snapshot.Skipped));
        return target;
    }

    private string CreateIncrementalBackup(CancellationToken cancellationToken)
    {
        var latest = FindLatestBackup();
        if (string.IsNullOrWhiteSpace(latest))
        {
            return CreateFullBackup(cancellationToken);
        }

        ManifestModel previousManifest;
        try
        {
            using var latestArchive = ZipFile.OpenRead(latest);
            previousManifest = ReadManifest(latestArchive);
        }
        catch
        {
            return CreateFullBackup(cancellationToken);
        }

        var previousIndex = previousManifest.FileIndex.ToDictionary(entry => entry.Path, StringComparer.OrdinalIgnoreCase);
        var snapshot = ComputeFileSnapshot(cancellationToken);
        var currentIndex = snapshot.Index.ToDictionary(entry => entry.Path, StringComparer.OrdinalIgnoreCase);
        var changedPaths = currentIndex
            .Where(pair => !previousIndex.TryGetValue(pair.Key, out var previous) || !string.Equals(previous.Sha256, pair.Value.Sha256, StringComparison.OrdinalIgnoreCase))
            .Select(pair => pair.Key)
            .ToHashSet(StringComparer.OrdinalIgnoreCase);
        var deletedPaths = previousIndex.Keys.Where(path => !currentIndex.ContainsKey(path)).OrderBy(path => path, StringComparer.OrdinalIgnoreCase).ToList();

        var tempPath = Path.Combine(_backupDirectory.FullName, $"{Guid.NewGuid():N}.bak");
        using (var sourceArchive = ZipFile.OpenRead(latest))
        using (var targetArchive = ZipFile.Open(tempPath, ZipArchiveMode.Create))
        {
            CopyArchiveEntries(sourceArchive, targetArchive, changedPaths, cancellationToken);
            WriteSnapshotEntries(targetArchive, snapshot.Map, changedPaths, cancellationToken);
            var manifest = BuildManifest(snapshot.Index, "incremental", Path.GetFileName(latest), deletedPaths);
            WriteTextEntry(targetArchive, "manifest.json", JsonSerializer.Serialize(manifest, JsonOptions));
            WriteTextEntry(targetArchive, "meta/log.txt", BuildLog(snapshot.Skipped));
        }

        File.Copy(tempPath, latest, overwrite: true);
        File.Delete(tempPath);
        return latest;
    }

    private Snapshot ComputeFileSnapshot(CancellationToken cancellationToken)
    {
        var index = new List<FileIndexEntry>();
        var map = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var skipped = new List<string>();

        foreach (var (archivePath, sourcePath) in ScanFiles())
        {
            cancellationToken.ThrowIfCancellationRequested();
            try
            {
                var fileInfo = new FileInfo(sourcePath);
                var hash = ComputeSha256(sourcePath);
                index.Add(new FileIndexEntry(archivePath, fileInfo.Length, (fileInfo.LastWriteTimeUtc - DateTime.UnixEpoch).TotalSeconds, hash));
                map[archivePath] = sourcePath;
            }
            catch (UnauthorizedAccessException)
            {
                skipped.Add($"Permission denied: {sourcePath}");
            }
            catch (Exception exc)
            {
                skipped.Add($"Skipped {sourcePath}: {exc.Message}");
            }
        }

        index.Sort((left, right) => StringComparer.OrdinalIgnoreCase.Compare(left.Path, right.Path));
        return new Snapshot(index, map, skipped);
    }

    private IEnumerable<(string ArchivePath, string SourcePath)> ScanFiles()
    {
        foreach (var pair in RootMutableFiles)
        {
            var sourcePath = Path.Combine(_configDirectory.FullName, pair.Key);
            if (File.Exists(sourcePath) && IsIncluded(Path.GetFileName(sourcePath)))
            {
                yield return (pair.Value, sourcePath);
            }
        }

        if (!_dataDirectory.Exists)
        {
            yield break;
        }

        foreach (var path in _dataDirectory.EnumerateFiles("*", SearchOption.AllDirectories).OrderBy(file => file.FullName, StringComparer.OrdinalIgnoreCase))
        {
            var relative = Path.GetRelativePath(_dataDirectory.FullName, path.FullName).Replace('\\', '/');
            if (IsIncluded(relative))
            {
                yield return ($"data/{relative}", path.FullName);
            }
        }
    }

    private static void CopyArchiveEntries(ZipArchive sourceArchive, ZipArchive targetArchive, ISet<string> changedPaths, CancellationToken cancellationToken)
    {
        var skipped = new HashSet<string>(changedPaths, StringComparer.OrdinalIgnoreCase) { "manifest.json", "meta/log.txt" };
        foreach (var entry in sourceArchive.Entries)
        {
            cancellationToken.ThrowIfCancellationRequested();
            if (skipped.Contains(entry.FullName) || string.IsNullOrWhiteSpace(entry.Name))
            {
                continue;
            }

            var targetEntry = targetArchive.CreateEntry(entry.FullName, CompressionLevel.Optimal);
            using var input = entry.Open();
            using var output = targetEntry.Open();
            input.CopyTo(output);
        }
    }

    private static void WriteSnapshotEntries(ZipArchive archive, IReadOnlyDictionary<string, string> fileMap, IEnumerable<string> paths, CancellationToken cancellationToken)
    {
        foreach (var archivePath in paths.OrderBy(path => path, StringComparer.OrdinalIgnoreCase))
        {
            cancellationToken.ThrowIfCancellationRequested();
            if (!fileMap.TryGetValue(archivePath, out var sourcePath) || !File.Exists(sourcePath))
            {
                continue;
            }

            archive.CreateEntryFromFile(sourcePath, archivePath, CompressionLevel.Optimal);
        }
    }

    private static void WriteTextEntry(ZipArchive archive, string path, string content)
    {
        var entry = archive.CreateEntry(path, CompressionLevel.Optimal);
        using var writer = new StreamWriter(entry.Open());
        writer.Write(content);
    }

    private ManifestModel BuildManifest(IReadOnlyList<FileIndexEntry> index, string mode, string? baseBackup, IReadOnlyList<string> deletedPaths)
        => new(
            BackupVersion,
            DateTimeOffset.UtcNow,
            mode,
            _appName,
            new SourcePaths(_configDirectory.FullName, _dataDirectory.FullName),
            index,
            deletedPaths,
            baseBackup);

    private static ManifestModel ReadManifest(ZipArchive archive)
    {
        var entry = archive.GetEntry("manifest.json") ?? throw new InvalidDataException("manifest.json fehlt im Backup.");
        using var reader = new StreamReader(entry.Open());
        return JsonSerializer.Deserialize<ManifestModel>(reader.ReadToEnd(), JsonOptions) ?? throw new InvalidDataException("manifest.json ist ungültig.");
    }

    private static string? ResolveRestorePath(string archiveMember, string targetDataDirectory, string targetConfigDirectory)
    {
        var normalized = archiveMember.Replace('\\', '/');
        if (string.Equals(normalized, "config/settings.json", StringComparison.OrdinalIgnoreCase))
        {
            return Path.Combine(targetConfigDirectory, "settings.json");
        }

        if (normalized.StartsWith("data_root/", StringComparison.OrdinalIgnoreCase))
        {
            return Path.Combine(targetConfigDirectory, normalized["data_root/".Length..].Replace('/', Path.DirectorySeparatorChar));
        }

        if (normalized.StartsWith("data/", StringComparison.OrdinalIgnoreCase))
        {
            return Path.Combine(targetDataDirectory, normalized["data/".Length..].Replace('/', Path.DirectorySeparatorChar));
        }

        return null;
    }

    public static string ClassifyArchiveMember(string member)
    {
        var path = (member ?? string.Empty).Replace('\\', '/');
        if (string.Equals(path, "config/settings.json", StringComparison.OrdinalIgnoreCase)) return "settings";
        if (string.Equals(path, "data_root/pins.json", StringComparison.OrdinalIgnoreCase)) return "orders";
        if (string.Equals(path, "data_root/tours.json", StringComparison.OrdinalIgnoreCase)) return "tours";
        if (string.Equals(path, "data_root/geocode_cache.json", StringComparison.OrdinalIgnoreCase)
            || string.Equals(path, "data_root/config.json", StringComparison.OrdinalIgnoreCase)) return "misc";
        if (string.Equals(path, "data/employees.json", StringComparison.OrdinalIgnoreCase)) return "employees";
        if (string.Equals(path, "data/vehicles.json", StringComparison.OrdinalIgnoreCase)) return "vehicles";
        return path.StartsWith("data/", StringComparison.OrdinalIgnoreCase) ? "other_data" : "misc";
    }

    private static string ComputeSha256(string path)
    {
        using var stream = File.OpenRead(path);
        using var sha = SHA256.Create();
        return Convert.ToHexString(sha.ComputeHash(stream));
    }

    private static bool IsIncluded(string relativePath)
        => !ExcludeGlobs.Any(pattern => MatchesGlob(relativePath, pattern));

    private static bool MatchesGlob(string path, string pattern)
    {
        var regex = "^" + System.Text.RegularExpressions.Regex.Escape(pattern)
            .Replace(@"\*", ".*")
            .Replace(@"\?", ".") + "$";
        return System.Text.RegularExpressions.Regex.IsMatch(path, regex, System.Text.RegularExpressions.RegexOptions.IgnoreCase);
    }

    private static string BuildLog(IReadOnlyList<string> skipped)
    {
        var lines = new List<string> { $"Backup created at {DateTimeOffset.UtcNow:O}" };
        lines.AddRange(skipped);
        return string.Join(Environment.NewLine, lines) + Environment.NewLine;
    }

    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    private sealed record Snapshot(IReadOnlyList<FileIndexEntry> Index, IReadOnlyDictionary<string, string> Map, IReadOnlyList<string> Skipped);

    private sealed record ManifestModel(
        [property: JsonPropertyName("backup_version")] int BackupVersion,
        [property: JsonPropertyName("created_at_iso")] DateTimeOffset CreatedAtIso,
        [property: JsonPropertyName("backup_type")] string BackupType,
        [property: JsonPropertyName("app_name")] string AppName,
        [property: JsonPropertyName("source_paths")] SourcePaths SourcePaths,
        [property: JsonPropertyName("file_index")] IReadOnlyList<FileIndexEntry> FileIndex,
        [property: JsonPropertyName("deleted_paths")] IReadOnlyList<string> DeletedPaths,
        [property: JsonPropertyName("incremental_base")] string? IncrementalBase);

    private sealed record SourcePaths(
        [property: JsonPropertyName("config_dir")] string ConfigDirectory,
        [property: JsonPropertyName("data_dir")] string DataDirectory);

    private sealed record FileIndexEntry(
        [property: JsonPropertyName("path")] string Path,
        [property: JsonPropertyName("size")] long Size,
        [property: JsonPropertyName("mtime")] double Mtime,
        [property: JsonPropertyName("sha256")] string Sha256);
}
